src/core/service.c

```diff
@@ -5035,6 +5100,134 @@ static int service_can_clean(Unit *u, ExecCleanMask *ret) {
         return 0;
 }
 
+static int service_live_mount(Unit *u,
+                const char *src,
+                const char *dst,
+                sd_bus_message *message,
+                MountInNamespaceFlags flags,
+                const MountOptions *options) {
+
+        _cleanup_(pidref_done) PidRef worker = PIDREF_NULL;
+        Service *s = ASSERT_PTR(SERVICE(u));
+        const char *propagate_directory;
+        PidRef *unit_pid;
+        int r;
+
+        assert(u);
+        assert(src);
+        assert(dst);
+        assert(message);
+
+        if (s->state != SERVICE_RUNNING)
+                return log_unit_warning_errno(u, SYNTHETIC_ERRNO(EBUSY), "Service is not running");
+
+        unit_pid = unit_main_pid(u);
+        if (!pidref_is_set(unit_pid))
+                return log_unit_warning_errno(u, SYNTHETIC_ERRNO(EBUSY), "Service is not running");
+
+        service_unwatch_control_pid(s);
+        s->mount_result = SERVICE_SUCCESS;
+        s->control_command = NULL;
+        s->control_command_id = _SERVICE_EXEC_COMMAND_INVALID;
+
+        r = service_arm_timer(s, /* relative= */ true, s->timeout_start_usec);
+        if (r < 0) {
+                log_unit_warning_errno(u, r, "Failed to install timer: %m");
+                goto fail;
+        }
+
+        propagate_directory = strjoina("/run/systemd/propagate/", u->id);
+
+        /* Given we are running from PID1, avoid doing potentially heavy I/O operations like opening images
+         * directly, and instead fork a worker process. We record the D-Bus message, so that we can reply
+         * after the operation has finished. This way callers can wait on the message and know that the new
+         * resource is available (or the operation failed) once they receive the response. */
+        r = unit_fork_helper_process(u, "(sd-mount-in-ns)", /* into_cgroup= */ false, &worker);
+        if (r < 0) {
+                log_unit_warning_errno(
+                                u,
+                                r,
+                                "Failed to fork process to mount '%s' on '%s' in unit's namespace: %m",
+                                src,
+                                dst);
+                goto fail;
+        }
+        if (r == 0) {
+                if (flags & MOUNT_IN_NAMESPACE_IS_IMAGE)
+                        r = mount_image_in_namespace(
+                                        unit_pid,
```

Comment:
hmm, one day we should not use the pid here btw, but instead send an fd to the ns from the relevant executor to pid 1, and then pin it there, and only go via that.

after all the main process might do its own ns stuff, chrooting and what not, and we are interfering with that. if otoh we operate on the original ns we create things should be much nicer, since the rearrangements we do will propagate according to mount propagation rules into anything the service does on its own. This in particular matters in case the service creates multiple namespaces for its service it then runs different subporcess in.

Reply:
(but this is for later, not asking for this to be fixed here and now)

Reply:
Sounds sensible, I've pushed a new item to the TODO file